### PR TITLE
Support regression trees in DecisionTreeTrainer

### PR DESCRIPTION
**Pull request by yolken**
- Modify existing DecisionTreeTrainer to support both classification- and regression-type splits; see http://www.stat.cmu.edu/~cshalizi/350-2006/lecture-10.pdf for a good overview of the latter.
- Misc. cleanups to associated files

@hectorgon can you take a look? Thanks.

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala [hunk 1]
```diff
@@ -129,94 +167,172 @@ object DecisionTreeTrainer {
...
+    for (example <- examples) {
+      val response = BoostedStumpsModel.getStumpResponse(candidateOpt.get, example)
+      val labelValue = example.get(rankKey).asScala.head._2
+
+      // Using "online algorithm" for computing mean and sum-squared errors as
+      // described in https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+      if (response) {
+        rightCount += 1
+        val delta = labelValue - rightMean
+        rightMean += delta / rightCount
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala]: This is not the right way to compute the variance .. you should accumulate the sum and sum squared and only divide by the count at the very end outside of the loop

File: training/src/test/scala/com/airbnb/aerosolve/training/TrainingTestHelper.scala [hunk 2]
```diff
@@ -47,4 +44,22 @@ object TrainingTestHelper {
...
+  def makeRegressionExamples = {
+    val examples = ArrayBuffer[Example]()
+    val label = ArrayBuffer[Double]()
+    val rnd = new java.util.Random(1234)
+
+    for (i <- 0 until 200) {
+      val x = 4.0 * (2.0 * rnd.nextDouble() - 1.0)
+      val y = 4.0 * (2.0 * rnd.nextDouble() - 1.0)
+
+      val quadradtic = x * x - 2 * y * y - 0.5 * x + 0.2 * y
```
Comment (hectorgon) [inline training/src/test/scala/com/airbnb/aerosolve/training/TrainingTestHelper.scala]: Can we include a small flat region by adding max(1.0, quadratic) to see if the decision tree compresses that region with a small number of splits?

File: training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala (hunk 1 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala]: Clarified that this is the online variant which is more numerically stable (Welford)

**Comment by hectorgon** [comment]
LGTM

File: training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala (hunk 1 above)
Comment (yolken) [inline training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala]: Made the comment more precise and also made the associated unit test a little more complex to verify that it works correctly.

File: training/src/test/scala/com/airbnb/aerosolve/training/TrainingTestHelper.scala (hunk 2 above)
Comment (yolken) [inline training/src/test/scala/com/airbnb/aerosolve/training/TrainingTestHelper.scala]: Changed the curve to have some flat regions and added a test that the algorithm does the right thing in the associated regions.

**Comment by yolken** [comment]
@hectorgon addressed your feedback. Feel free to take another look.

**Pull request merged**